src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectWorkspaceStateGenerator.cs

```diff
@@ -4,133 +4,116 @@
 using Microsoft.CodeAnalysis.Razor.ProjectSystem;
 using Microsoft.CodeAnalysis.Razor.Workspaces;
 
 namespace Microsoft.VisualStudio.Razor;
 
 [Export(typeof(IProjectWorkspaceStateGenerator))]
 [method: ImportingConstructor]
-internal sealed class ProjectWorkspaceStateGenerator(
+internal sealed partial class ProjectWorkspaceStateGenerator(
     IProjectSnapshotManager projectManager,
     ITagHelperResolver tagHelperResolver,
-    ProjectSnapshotManagerDispatcher dispatcher,
-    IErrorReporter errorReporter,
+    ILoggerFactory loggerFactory,
     ITelemetryReporter telemetryReporter)
     : IProjectWorkspaceStateGenerator, IDisposable
 {
-    // Internal for testing
-    internal readonly Dictionary<ProjectKey, UpdateItem> Updates = new();
-
     private readonly IProjectSnapshotManager _projectManager = projectManager;
     private readonly ITagHelperResolver _tagHelperResolver = tagHelperResolver;
-    private readonly ProjectSnapshotManagerDispatcher _dispatcher = dispatcher;
-    private readonly IErrorReporter _errorReporter = errorReporter;
+    private readonly ILogger _logger = loggerFactory.GetOrCreateLogger<ProjectWorkspaceStateGenerator>();
     private readonly ITelemetryReporter _telemetryReporter = telemetryReporter;
-    private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(initialCount: 1);
-
-    private bool _disposed;
+    private readonly SemaphoreSlim _semaphore = new(initialCount: 1);
 
-    // Used in unit tests to ensure we can control when background work starts.
-    public ManualResetEventSlim? BlockBackgroundWorkStart { get; set; }
+    private readonly CancellationTokenSource _disposeTokenSource = new();
+    private readonly Dictionary<ProjectKey, UpdateItem> _updates = [];
 
-    // Used in unit tests to ensure we can know when background work finishes.
-    public ManualResetEventSlim? NotifyBackgroundWorkCompleted { get; set; }
+    private ManualResetEventSlim? _blockBackgroundWorkStart;
+    private ManualResetEventSlim? _notifyBackgroundWorkCompleted;
 
-    public Task UpdateAsync(Project? workspaceProject, IProjectSnapshot projectSnapshot, CancellationToken cancellationToken)
+    public void Dispose()
     {
-        if (projectSnapshot is null)
+        if (_disposeTokenSource.IsCancellationRequested)
         {
-            throw new ArgumentNullException(nameof(projectSnapshot));
+            return;
         }
 
-        return _dispatcher.RunAsync(
-            () =>
-            {
-                if (_disposed)
-                {
-                    return;
-                }
+        _disposeTokenSource.Cancel();
+        _disposeTokenSource.Dispose();
 
-                if (Updates.TryGetValue(projectSnapshot.Key, out var updateItem) &&
-                    !updateItem.Task.IsCompleted &&
-                    !updateItem.Cts.IsCancellationRequested)
-                {
-                    updateItem.Cts.Cancel();
-                }
+        foreach (var (_, updateItem) in _updates)
```

Comment:
is there a possible race with dispose here?

Reply:
Yes, that's a race. Will fix. Thanks!